File: 97-interleaving-string/97-interleaving-string.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        n1,n2,n3 = len(s1), len(s2), len(s3)
        if n3 != n1 + n2:
            return False
        record = dict()
        def find(s1,s2,s3):
            if (len(s1) == 0 and s2 != s3) or (len(s2) == 0 and s1 != s3):
                record[(s1,s2,s3)] = False
                return False
            elif (len(s1) == 0 and s2 == s3) or (len(s2) == 0 and s1 == s3):
                record[(s1,s2,s3)] = True
                return True
            if (s1,s2,s3) in record:
                return record[(s1,s2,s3)]
            if s1[0] == s3[0] and s2[0] == s3[0]:
                res = (find(s1[1:],s2,s3[1:]) or find(s1,s2[1:],s3[1:]))
                record[(s1,s2,s3)] = res
                return res
            elif s1[0] == s3[0]:
                res = find(s1[1:],s2,s3[1:])
                record[(s1,s2,s3)] = res
                return res
            elif s2[0] == s3[0]:
                res = find(s1,s2[1:],s3[1:])
                record[(s1,s2,s3)] = res
                return res
            else:
                record[(s1,s2,s3)] = False
                return False
        return find(s1,s2,s3)
        

File: 97-interleaving-string/97-interleaving-string.java
class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int n = s1.length(), m = s2.length();
        boolean[][] dp = new boolean[n + 1][m + 1];
        if(n + m != s3.length())return false;
        
        dp[0][0] = true;
        char s1Char;
        char s2Char;
        char target;
        for(int i = 0; i < n + 1; i++)
            for(int j = 0; j < m + 1; j++){
                if(i == 0 && j == 0)continue;

                dp[i][j] = false;
                target = s3.charAt(i+j-1);
                if(i == 0){
                    s2Char = s2.charAt(j-1);
                    dp[i][j] = dp[i][j-1] &&s2Char == target;
                    continue;
                }
                s1Char = s1.charAt(i-1);
                if(j == 0){
                    dp[i][j] = dp[i-1][j] && s1Char == target;
                    continue;
                }
                s2Char = s2.charAt(j-1);
                if(s1Char == target){
                    dp[i][j]  |= dp[i-1][j];
                }
                if(s2Char == target){
                    dp[i][j] |= dp[i][j-1];
                }
             //   System.out.printf("%d %d\n", i, j);
           //     System.out.println(dp[i][j]);
            }
       // System.out.println(dp[2][0]);
        return dp[n][m];
        
    }
}